Made fractal algorithms more flexible so that colors can be customized

Lehonti opened the pull request:
There isn't a gradient control for use in `SimpleEffectDialog` yet, and the gradient type that was used is still undocumented, and is `internal` within `Pinta.Effects`. This is intentional, as we have to get it right before making it part of the public API.

... but until then we could add a combobox so that the user can choose among a few pre-defined color schemes. What do you think? 

Here's with the original colors (which is now the default gradient):
![fractal_default](https://github.com/PintaProject/Pinta/assets/17771375/057ab16b-7100-4410-bdc2-ad094b5d379d)

Here's with the same gradient stops but different colors:
![fractal_green_magenta](https://github.com/PintaProject/Pinta/assets/17771375/8bdce57f-cddb-48c8-8277-d34f0762c88e)
![fractal_brightannoying](https://github.com/PintaProject/Pinta/assets/17771375/f8e2164e-b02b-44fe-ad19-fb12e12950cf)

Here's with the original colors but different stops:
![mandelbrotdifferentstops](https://github.com/PintaProject/Pinta/assets/17771375/bff06fe3-e583-4e3b-9de6-b59ec03d12dd)

cameronwhite commented:
I haven't looked through the code in depth, but one idea could be to have the EffectData take a gradient parameter rather than an enum, so that the effect's API doesn't change if we later add support for a widget that can specify a custom gradient.
The SimpleEffectDialog could then handle how to create a gradient widget, e.g. just showing a preset combo box for now

Lehonti commented:
I understand that, but in order to do that we would have to clearly define the public interface of the gradient type, but since this isn't done yet what if we use an enum, and we make the enum-typed property and the gradient type `internal`? This way, the effect will be customizable within Pinta, but third parties won't notice the change in API until we make the property (and gradient type) `public`

cameronwhite commented:
Would this prevent writing a unit test since the property is `internal`?

I'm not sure if this is a high enough priority that it needs to be squeezed in right away though - why not take the time to implement a gradient API that you're happy with making public?

Lehonti commented:
I was thinking that the feature would increase the popularity of the project and maybe bring more contributors, but yeah, we could wait to get it right first.

I am almost happy with the methods in my gradient class; although it may need some refinements, like methods for adding new gradient stops and such, and adding some `static` creational methods so that the API doesn’t change even if some feature is added to the class. But what do you think about the public interface of the class as it is now?

As for the testing, I think we could do `[assembly: InternalsVisibleTo ("Pinta.Effects.Test")]` so that any piece of code can be tested even if it's not part of the public API.

cameronwhite commented:
👍 
I'm away for the next couple weeks so won't have time to review this for the next little while, sorry!

cameronwhite commented on Pinta.Core/Effects/ColorMapping.cs:
```diff
@@ -0,0 +1,61 @@
+
+namespace Pinta.Core;
+
+public abstract class ColorMapping
```
I'm not really clear on the split here between `ColorMapping` and `ColorGradient`, and if it's necessary to have the base class at all? In particular, having factory methods here which create a `ColorGradient` feels wrong for a base class

Lehonti replied:
There could be color mappings that are not gradients, hence this structure. But I could get rid of the `abstract` class

cameronwhite replied:
I don't necessarily mind having the separate `ColorMapping` to allow for other implementations, but if so I think things should be less tightly coupled.
e.g.
- Just make `ColorMapping` an interface and move anything else to `ColorGradient`
- Should the effects only work with a `ColorMapping` and leave it to the effect dialog to create a gradient?

cameronwhite commented on Pinta.Core/Effects/ColorGradient.cs:
```diff
@@ -0,0 +1,94 @@
+	{
+		if (position == MinimumPosition) return StartColor;
+		if (position == MaximumPosition) return EndColor;
+		if (!IsMapped (position)) throw new ArgumentOutOfRangeException (nameof (position));
```
If the position is out of range, perhaps this should just clamp to the start / end color? Otherwise this requires the caller to always carefully clamp which might be unexpected

Lehonti replied:
Do you think clamping would be better, or adjusting the start/end color to be an intermediate between the out-of-range color and the next stop?

cameronwhite replied:
I think just clamping makes sense, since there's nothing well-defined before the start color to continue blending into.
One other thing I just noticed too is that the effects currently assume that the gradient has a 0-1024 range, which might not actually be the case for whatever mapping the effect is provided. So that's also something to think about with the API, e.g. maybe it's easiest for the mapping to work in a 0-1 range and the effect remaps its 0-1024 value to that

cameronwhite commented on Pinta.Core/Effects/ColorGradient.cs:
```diff
@@ -0,0 +1,94 @@
+
+	public sealed override ColorBgra GetColor (double position)
+	{
+		if (position == MinimumPosition) return StartColor;
```
If the list of `sorted_stops` includes the start / end colours, that might unify some of these branches and simplify the logic a bit?

cameronwhite commented on Pinta.Core/Effects/ColorGradient.cs:
```diff
@@ -0,0 +1,94 @@
+
+	/// <returns>Index of number immediately lower than target. If not found, it returns -1</returns>
+	/// <param name="arr">Array assumed to be sorted by key</param>
+	private static int BinarySearchLowerOrEqual (IReadOnlyList<GradientStop> arr, double target)
```
Would https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.binarysearch?view=net-7.0 be usable here?

Lehonti replied:
The method could be modified to use what you are mentioning, but a simple call to it wouldn't be sufficient, because _exact_ matches are not sufficient. We should be able to search for the _closest_ match in either direction.

Quoting the documentation:

---

Returns
[Int32](https://learn.microsoft.com/en-us/dotnet/api/system.int32?view=net-7.0)

The zero-based index of item in the sorted [List<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1?view=net-7.0), if item is found; otherwise, a negative number that is the bitwise complement of the index of the next element that is larger than item or, if there is no larger element, the bitwise complement of [Count](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.count?view=net-7.0#system-collections-generic-list-1-count).

---

I would still have to test if the returned index is in range or if it's a bitwise complement of the next element that is larger, and so on, so it would result in several lines of code one way or the other.

cameronwhite replied:
Yeah it's definitely not just a one-liner, but IMO building on top of the standard library algorithms is better vs rolling our own similar code

Lehonti replied:
I just changed the binary search so that it uses standard library methods, too

Lehonti commented:
Now I've implemented the most important of your suggestions. The default gradients go from 0 to 1 (and I added a `Resized()` method to allow us to adjust the gradient to the values we need), and there is no base `ColorMapping class`.

cameronwhite commented on Pinta.Core/Effects/ColorGradient.cs:
```diff
@@ -0,0 +1,183 @@
+
+	private ColorBgra HandleNoStops (double position)
+	{
+		double valueSpan = MaximumPosition - MinimumPosition;
```
This code here could perhaps be factored out a method like `Utility.InvLerp()` and reused in `HandleWithStops` below - it's doing the inverse of `Lerp(MinimumPosition, MaximumPosition, fraction)`

Lehonti replied:
I just created and commited an `InvLerp()` method

cameronwhite commented on Pinta.Gui.Widgets/Utilities/GradientHelper.cs:
```diff
@@ -0,0 +1,122 @@
+	[Caption ("Piña Colada")]
+	PinaColada,
+
+	[Caption ("Sakura Sigh")]
```
Was there a standard list of gradients this was sourced from? I'm just wondering in terms of how to describe what these strings mean for translators

Lehonti replied:
No, I made them up. The gradients are not based on anything in particular, either; I just created several different ones so that there is some variety in the gradients that the user can choose.

cameronwhite replied:
OK 👍 
To make this easier for translators, you could add comments like `// Translators: This is the name of a gradient preset which blends from yellow to green` or whatever makes sense to explain it. Comments like that are picked up when the source code is parsed for translatable strings, and made visible on Weblate for translators

I'm not entirely sure how easy it'll be to translate a few of the labels, e.g. `La Bella Italia` and `Sakura Sigh`?

Lehonti replied:
No problem, I just changed a couple of names to make them easier to translate, and I also left a few comments for translators (but could you tell me if I did it right?)

cameronwhite replied:
yep, I think that looks correct 👍

cameronwhite commented on Pinta.Core/Effects/ColorGradient.cs:
```diff
@@ -0,0 +1,183 @@
+	private static double DefaultMaximumValue () => 255;
+	private static IEnumerable<KeyValuePair<double, ColorBgra>> EmptyStops () => Enumerable.Empty<KeyValuePair<double, ColorBgra>> ();
+
+	public static ColorGradient Create ()
```
Is this empty Create() method useful for any specific purpose?

Lehonti replied:
It's just the default behavior if nothing is specified. In any case, I've just removed that one and other method that wasn't used, in order to keep the code slim.

cameronwhite commented on Pinta.Effects/Effects/JuliaFractalEffect.cs:
```diff
@@ -147,6 +158,9 @@ public sealed class JuliaFractalData : EffectData
 		public int Zoom { get; set; } = 1;
 
+		[Caption ("Color Scheme")]
+		public PredefinedGradients ColorScheme { get; set; } = PredefinedGradients.Bonfire;
```
Is it possible to have the effect data just contain a `ColorGradient`, and leave it up to `SimpleEffectDialog` to render that as a list of the pre-defined gradients for now?

Lehonti replied:
It is possible, and that was the original idea, but I did it this way because it doesn't require changes in the public API.

As you've mentioned before, there are core libraries that should be sufficient for the user of the library to write new effects, and then there is `Pinta.Effects`. I think that if everything is contained in `Pinta.Effects` we can play fast and loose with the effects until we get the API just right (thus making it easier to write new effects more quickly, without worrying about the API until we get it right).

In this case, the core API is sufficient because `SimpleEffectDialog` can handle `enum` types.

But this is, of course, an opinion. What is your view, given the above?

cameronwhite replied:
OK, yeah I think keeping everything in Pinta.Effects is a reasonable approach for now, then

cameronwhite reviewed (commented):
I think this is looking pretty good so far 👍 
`ColorGradient` should also have doc comments added to it once finalized

Lehonti commented:
What (besides the `InvLerp` method) is missing?

cameronwhite commented:
> What (besides the `InvLerp` method) is missing?

I think just that, and documenting any remaining public methods like `Resized()`
There also was a suggestion before of using standard library methods instead of the manual binary search - any thoughts on that?

Lehonti commented:
There seems to be some bug that is making the tests fail, even though the fractals look visually OK. I'll be looking into it

Lehonti commented:
The problem was a comparison that was supposed to be _greater or equal_, but was only _greater_. I think it's ready for merging now.

cameronwhite commented:
Thank you! I think the changes look good 👍 
After taking a look over the unit tests, one final question is whether `GetRandomColor()` is necessary or if we can just use some arbitrary fixed colours like black or white. Having random values in a test makes me nervous in general :)

Lehonti commented:
It's understandable (and reasonable) that you want the tests to be deterministic and reproducible.

`GetRandomColor()` is not necessary as such, but it's more of a way to emphasize that a certain result holds no matter what color is chosen. Of course, it need not be non-deterministic; we can find some other way to test that.

In any case, I've now made the changes.

cameronwhite commented:
Looks good! Thank you for working on this
